src/cmt_filter.c

```diff
@@ -0,0 +1,316 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  CMetrics
+ *  ========
+ *  Copyright 2021-2024 The CMetrics Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <cmetrics/cmetrics.h>
+#include <cmetrics/cmt_log.h>
+#include <cmetrics/cmt_map.h>
+#include <cmetrics/cmt_counter.h>
+#include <cmetrics/cmt_gauge.h>
+#include <cmetrics/cmt_untyped.h>
+#include <cmetrics/cmt_histogram.h>
+#include <cmetrics/cmt_summary.h>
+#include <cmetrics/cmt_filter.h>
+
+static int compare_label_keys(struct cmt_map *src, const char *label_key,
+                              void *compare_ctx, int (*compare)(void *compare_ctx, const char *str, size_t slen),
+                              int flags)
+{
+    struct cfl_list *head;
+    struct cmt_map_label *label;
+    size_t label_key_size = 0;
+
+    if (label_key != NULL) {
```

Comment:
if there is a chance this can be NULL the code below will not work. Either the check is not necessary or we just return with -1 if is NULL

Reply:
There is a chance to become NULL. Because we are able to specify an external function only here.
And this should be injected via compare_ctx and callback function.

Reply:
label_key is only needed for prefix matching case. I added a clause which checks for CMT_FILTER_PREFIX  flag.